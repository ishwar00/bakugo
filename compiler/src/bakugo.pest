//! Parsing expression grammars(PEG) for subset of the Go programming language spec

// TODO:
// 1. Disallow NUL character(U+0000) in the source text
// 2. Ignore a UTF-8-encoded byte order mark (U+FEFF) if it is the first
// Unicode code point in the source text. A byte order mark may be disallowed
// anywhere else in the source.
// 3. consider comment containing newlines as newlines. A general comment containing no
// newlines acts like a space. Any other comment acts like a newline.
// 4. Semicolons
// 5. Keywords: restrict keywords as identifiers. Probably in AST.

// Questions:
//

/// Source Code Representation
/// Characters

/// escape sequence `\n`
NewLine     = { "\u{000A}" }
UnicodeChar = { !(NewLine) ~ ANY }
// refer:
// 1. https://pest.rs/book/grammars/built-ins.html
// 2. page 175 of https://www.unicode.org/versions/Unicode8.0.0/ch04.pdf
UnicodeLetter = { LETTER }
UnicodeDigit  = { DECIMAL_NUMBER }

/// Letters and digits

Letter       = { UnicodeLetter | "_" }
DecimalDigit = { '0'..'9' }
OctalDigit   = { '0'..'7' }
HexDigit     = { '0'..'9' | 'A'..'F' | 'a'..'f' }

/// Lexical elements

/// space (U+0020), horizontal tabs (U+0009), carriage returns (U+000D) and
/// newline (U+000A)
WHITESPACE = _{ "\u{0020}" | "\u{0009}" | "\u{000D}" | "\u{000A}" }

/// No comments containg `NewLine`.
// Refer TODO[3]
COMMENT = _{ "/*" ~ (!("*/") ~ ANY)* ~ "*/" }

/// Punctuation
Semicolon = { ";" }
Comma = { "," }

/// Identifier
Ident = @{ Letter ~ (Letter | UnicodeDigit)* }

IntLit        = @{ "0" | ('1'..'9' ~ ("_"? ~ DecimalDigits)?) }
FloatLit      = @{ DecimalDigits ~ "." ~ DecimalDigits? | "." ~ DecimalDigits }
DecimalDigits = @{ DecimalDigit ~ ("_"? ~ DecimalDigit)* }

/// Rune literals
RuneLit        = @{ "'" ~ (ByteValue | UnicodeValue) ~ "'" }
UnicodeValue   =  { LittleUValue | BigUValue | EscapedChar | UnicodeChar }
ByteValue      =  { HexByteValue | OctalByteValue }
OctalByteValue = @{ "\\" ~ OctalDigit{3} }
HexByteValue   = @{ "\\" ~ "x" ~ HexDigit{2} }
LittleUValue   = @{ "\\" ~ "u" ~ HexDigit{4} }
BigUValue      = @{ "\\" ~ "U" ~ HexDigit{8} }
EscapedChar    = @{ "\\" ~ ("a" | "b" | "f" | "n" | "r" | "t" | "v" | "\\" | "'" | "\"") }

/// string literals
StringLit = { RawStringLit | InterpretedStringLit }
// `UnicodeChar | NewLine` is equivalent to `ANY``
RawStringLit         = @{ "`" ~ (!"`" ~ ANY)* ~ "`" }
InterpretedStringLit = @{ "\"" ~ (!"\"" ~ (ByteValue | UnicodeValue))* ~ "\"" }

// Types
Type     =  { TypeName | Tuple }
TypeName = _{ Ident }

// TODO: add it in spec
// Tuple
Tuple    = { "(" ~ (TypeList | (Type ~ ",")) ~ ")" }
TypeList = { Type ~ ("," ~ Type)+ }

// Function types
Signature      =  { Parameters ~ Result? }
Result         = _{ Type }
Parameters     =  { "(" ~ (ParameterList ~ ","?)? ~ ")" }
ParameterList  =  { ParameterDecl ~ ("," ~ ParameterDecl)* }
ParameterDecl  =  { (IdentifierList ~ Type) | (Type ~ ("," ~ Type)*) }
IdentifierList =  { Ident ~ ("," ~ Ident)* }

// Blocks
// *(a.method())
Block         = _{ "{" ~ StatementList ~ "}" }
StatementList =  { (Statement ~ Semicolon)* }

// Declarations
Declaration  = { ConstDecl | TypeDecl | VarDecl }
TopLevelDecl = { Declaration | FunctionDecl }

// Const declartions
ConstDecl      = { "const" ~ (ConstSpec | "(" ~ (ConstSpec ~ Semicolon)+ ~ ")") }
ConstSpec      = { IdentifierList ~ (Type? ~ "=" ~ ExpressionList)? }
ExpressionList = { Expression ~ (Comma ~ Expression)* }

// Type declarations
TypeDecl = { "type" ~ (TypeSpec | "(" ~ (TypeSpec ~ Semicolon)+ ~ ")") }
TypeSpec = _{ TypeDef }

// TODO: correct TypeDef spec (remove type parameters)
// Type definitions
TypeDef = { Ident ~ Type }

// Variable declaration
VarDecl = { "var" ~ (VarSpec | "(" ~ (VarSpec ~ Semicolon)+ ~ ")") }
VarSpec = { IdentifierList ~ ((Type ~ ("=" ~ ExpressionList)?) | ("=" ~ ExpressionList)) }

// Function declaration
FunctionDecl =  { "func" ~ FunctionName ~ Signature ~ FunctionBody }
FunctionName = _{ Ident }
FunctionBody = _{ Block }

// TODO: update the spec
// Operands
Operand     = _{ Literal | OperandName | "(" ~ Expression ~ ")" }
Literal     = _{ BasicLit }
BasicLit    = _{ FloatLit | IntLit | RuneLit | StringLit }
OperandName = _{ Ident }

// Primary expressions
// ref: https://www.csd.uwo.ca/~mmorenom/CS447/Lectures/Syntax.html/node8.html
PrimaryExpr      = _{ FunctionCall | Operand }
FunctionCall     = { Operand ~ PrimaryExprPrime }
PrimaryExprPrime = _{ (Arguments ~ PrimaryExprPrime?) }

Arguments = { "(" ~ ((ExpressionList | (Type ~ (Comma ~ ExpressionList)?)) ~ ","?)? ~ ")" }

// Operators
// ref: https://www.csd.uwo.ca/~mmorenom/CS447/Lectures/Syntax.html/node8.html
Expression      = _{ UnaryExpr ~ ExpressionPrime? }
ExpressionPrime = _{ BinaryOp ~ Expression ~ ExpressionPrime? }
UnaryExpr       = _{ (UnaryOp ~ UnaryExpr) | PrimaryExpr }

BinaryOp  = _{ RelOp | SumOp | ProductOp }

RelOp     = _{ Eq | NotEq | Le | Lt | Ge | Gt }
Eq        =  { "==" }
NotEq     =  { "!=" }
Le        =  { "<=" }
Lt        =  { "<" }
Ge        =  { ">=" }
Gt        =  { ">" }

SumOp     = _{ Plus | Minus | Pipe }
Plus      =  { "+" }
Minus     =  { "-" }
Pipe      =  { "|" }

ProductOp = _{ Star | Slash | Percent | And }
Star      =  { "*" }
Slash     =  { "/" }
Percent   =  { "%" }
And       =  { "&" }

UnaryOp    = _{ UnaryPlus | UnaryMinus | UnaryNot | UnaryStar | UnaryAnd }
UnaryPlus  =  { "+" }
UnaryMinus =  { "-" }
UnaryStar  =  { "*" }
UnaryAnd   =  { "&" }
UnaryNot   =  { "!" }

// Statements

Statement = _{ ReturnStmt | Declaration | SimpleStmt | Block | IfStmt }

SimpleStmt = _{ ExpressionStmt | IncDecStmt | Assignment }

ExpressionStmt = { Expression }

IncDecStmt = { Expression ~ ("++" | "--") }

Assignment = { ExpressionList ~ "=" ~ ExpressionList }

IfStmt = { "if" ~ (SimpleStmt ~ Semicolon)? ~ Expression ~ Block ~ ("else" ~ (IfStmt | Block))? }

ReturnStmt = { "return" ~ ExpressionList? }

SourceFile = { SOI ~ (TopLevelDecl ~ Semicolon)* ~ EOI }
