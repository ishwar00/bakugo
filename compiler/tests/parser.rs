use std::{fmt::Display, fs, path::Path};

use bakugo::{ast::BakugoParsingErrorDisplay, parser::*};
use insta::{
    assert_debug_snapshot, assert_display_snapshot, assert_snapshot, assert_yaml_snapshot, glob,
};
use miette::{GraphicalReportHandler, GraphicalTheme, NamedSource};
use pest::Parser;

struct BakugoParsingErrDebug(BakugoParsingErrorDisplay);

impl Display for BakugoParsingErrDebug {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let report_handler = GraphicalReportHandler::new_themed(GraphicalTheme::unicode_nocolor());
        report_handler.render_report(f, &self.0)
    }
}

#[test]
fn test_parser() {
    fn test_parser_from_path(path: &Path) {
        let input = fs::read_to_string(path).unwrap();
        let parsed = parse_string(&input);
        match parsed {
            Ok(parsed) => assert_yaml_snapshot!(parsed),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    }

    glob!("examples/*.bakugo", test_parser_from_path);
    glob!("examples/ast_error/*.bakugo", test_parser_from_path);
    glob!("examples/parser_error/*.bakugo", test_parser_from_path);
}

#[test]
fn test_ast() {
    fn test_ast_from_path(path: &Path) {
        let input = fs::read_to_string(path).unwrap();
        let mut parsed = parse_string(&input).unwrap();
        let source_file = parsed.next().unwrap();
        let source = NamedSource::new(path.file_name().unwrap().to_str().unwrap(), input.clone());
        match construct_ast(source, source_file) {
            Ok(parsed) => assert_debug_snapshot!(parsed),
            Err(err) => assert_display_snapshot!(BakugoParsingErrDebug(err)),
        }
    }

    glob!("examples/*.bakugo", test_ast_from_path);
    glob!("examples/ast_error/*.bakugo", test_ast_from_path);
}

#[test]
fn test_identifier() {
    let ident_tests = vec![
        "a",
        "_x9",
        "ThisIsVariable",
        "αβ",
        "_",
        // invalid
        "9_nope",
        "(_)43",
        " space",
    ];

    for ident in ident_tests {
        let parse_result = BakugoParser::parse(Rule::Ident, ident);
        match parse_result {
            Ok(parsed) => assert_yaml_snapshot!(parsed.as_str()),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    }
}

#[test]
fn test_decimal_lit() {
    let decimal_tests = vec![
        "433",
        "3_4_3",
        "0",
        "170141183460469231731687303715884105727",
        "170_141183_460469_231731_687303_715884_105727",
        // invalid
        "_43",
        "i32",
        " space",
        "(333)",
    ];

    for decimal_test in decimal_tests {
        let parse_result = BakugoParser::parse(Rule::IntLit, decimal_test);
        match parse_result {
            Ok(parsed) => assert_yaml_snapshot!(parsed.as_str(), decimal_test),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    }
}

#[test]
fn test_rune_lit() {
    let runes = r"
        'a'
        'ä'
        '本'
        '\t'
        '\000'
        '\007'
        '\377'
        '\x07'
        '\xff'
        '\u12e4'
        '\U00101234'
        '\''

        'aa'
        '\k'
        '\xa'
        '\0'
        '\400'
        '\uDFFF'
        '\U00110000'
    ";
    // NOTE:
    // '\''        // rune literal containing single quote character
    //'aa'         // illegal: too many characters
    //'\k'         // illegal: k is not recognized after a backslash
    //'\xa'        // illegal: too few hexadecimal digits
    //'\0'         // illegal: too few octal digits
    //'\400'       // illegal: octal value over 255
    //'\uDFFF'     // illegal: surrogate half
    //'\U00110000' // illegal: invalid Unicode code point

    let rune_tests: Vec<&str> = runes
        .lines()
        .map(|line| line.trim())
        .filter(|trim_line| !trim_line.is_empty())
        .collect();

    for rune_test in rune_tests {
        let parse_result = BakugoParser::parse(Rule::RuneLit, rune_test);
        match parse_result {
            Ok(parsed) => assert_yaml_snapshot!(parsed.as_str(), rune_test),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    }
}

#[test]
fn test_string_lit() {
    let strings = [
        r#" `abc` "#,
        r#" "\n" "#,
        r#" "\"" "#,
        r#" "Hello, world!\n" "#,
        r#" "日本語" "#,
        r#" `日本語` "#,
        r#" "\u65e5本\U00008a9e" "#,
        r#" "\U000065e5\U0000672c\U00008a9e" "#,
        r#" "\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e" "#,
        r#" "\xff\u00FF" "#,
        r#" "\uD800" "#,
        r#" "\U00110000" "#,
    ];

    let string_tests: Vec<&str> = strings.iter().map(|string| string.trim()).collect();

    for string_test in string_tests {
        let parse_result = BakugoParser::parse(Rule::StringLit, string_test);
        match parse_result {
            Ok(mut parsed) => {
                let string_lit = parsed.next().unwrap();
                if let Rule::RawStringLit | Rule::InterpretedStringLit = string_lit.as_rule() {
                    assert_yaml_snapshot!(string_lit.as_str(), string_test);
                } else {
                    panic!(
                        "testing error: recieved {}, with rule {:?}",
                        string_test,
                        string_lit.as_rule()
                    );
                }
            }
            Err(err) => assert_snapshot!(err.to_string()),
        }
    }
}
