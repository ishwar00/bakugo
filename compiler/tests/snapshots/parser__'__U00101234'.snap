---
source: tests/parser.rs
expression: rune_test
---
"'\\U00101234'"

